File: Trie/trie.cpp
#include <iostream>
#include <unordered_map>
using namespace std;

class Node
{
    public:
    char data;
    unordered_map<char, Node*> children;
    bool terminal;

    Node(char data) {
        this->data = data;
        terminal = false;
    }
};

class Trie
{
    Node *root;
    int count;
    public:
    Trie() {
        root = new Node('\0');
        count = 0;
    }
    void insert(char *w) {
        Node *temp = root;
        for (int i = 0; w[i] != '\0'; i++) {
            char ch = w[i];
            if (temp->children.count(ch)) {
                temp = temp->children[ch];
            }
            else {
                Node* n = new Node(ch);
                temp->children[ch] = n;
                temp = n;
            }
        }
        temp->terminal = true;
    }
    bool find(char *w) {
        Node* temp = root;
        for (int i = 0; w[i] != '\0'; i++) {
            char ch = w[i];
            if (temp->children.count(ch) == 0) {
                return false;
            }
            else {
                temp = temp->children[ch];
            }
        }
        return temp->terminal;
    }
};
int main()
{
    Trie t;
    char words[][15] = {"a", "hello", "not", "news", "apple"};
    char w[10];
    cout << "Enter a word to search!" << '\n';
    cin >> w;
    for (int i = 0; i < 5; i++) {
        t.insert(words[i]);
    }
    if (t.find(w)) {
        cout << "Word is present!" << '\n';
    }
    else cout << "Word is not present" << '\n';

    return 0;
}

File: Trie/Trie.java
import java.lang.Override;

public class Trie {
  public static int ALPHABET_SIZE = 26;

  private class Node {
    private char value;
    private Node[] children;
    private boolean isEndOfWord;

    public Node(char value) {
      this.value = value;
      this.children = new Node[ALPHABET_SIZE];
      this.isEndOfWord = false;
    }

    @Override
    public String toString() {
      return "value = " + value;
    }
  }

  private Node root;

  public Trie() {
    root = new Node('\0');
  }

  public void insert(String word) {
    var current = root;
    for (var ch : word.toCharArray()) {
      // handles lowercase letters right now
      var index = ch - 'a';
      if (current.children[index] == null) {
        current.children[index] = new Node(ch);
      }
      current = current.children[index];
    }
    current.isEndOfWord = true;
  }
}
